zkevm-circuits/src/anchor_tx_circuit.rs

```diff
@@ -0,0 +1,544 @@
+//! Anchor circuit implementation.
+
+#[cfg(any(feature = "test", test, feature = "test-circuits"))]
+mod dev;
+#[cfg(any(feature = "test", test, feature = "test-circuits"))]
+pub use dev::TestAnchorTxCircuit;
+mod sign_verify;
+#[cfg(any(feature = "test", test))]
+mod test;
+
+use crate::{
+    evm_circuit::util::constraint_builder::{BaseConstraintBuilder, ConstrainBuilderCommon},
+    table::{LookupTable, PiFieldTag, PiTable, TxContextFieldTag, TxFieldTag, TxTable},
+    tx_circuit::TX_LEN,
+    util::{Challenges, SubCircuit, SubCircuitConfig},
+    witness::{self, Taiko, Transaction},
+};
+use eth_types::{Field, ToScalar};
+use gadgets::util::{select, Expr};
+use halo2_proofs::{
+    circuit::{Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, SecondPhase, Selector},
+    poly::Rotation,
+};
+use sign_verify::SignVerifyConfig;
+use std::marker::PhantomData;
+
+use self::sign_verify::GOLDEN_TOUCH_ADDRESS;
+
+// The first of txlist is the anchor tx
+const ANCHOR_TX_ID: usize = 1;
+const ANCHOR_TX_VALUE: u64 = 0;
+const ANCHOR_TX_IS_CREATE: bool = false;
+const ANCHOR_TX_GAS_PRICE: u64 = 0;
+// TODO: calculate the method_signature
+const ANCHOR_TX_METHOD_SIGNATURE: u32 = 0;
+const MAX_DEGREE: usize = 9;
+const BYTE_POW_BASE: u64 = 1 << 8;
+
+// anchor(bytes32,bytes32,uint64,uint64) = method_signature(4B)+1st(32B)+2nd(32B)+3rd(8B)+4th(8B)
+const ANCHOR_CALL_DATA_LEN: usize = 84;
+
+struct CallData {
+    start: usize,
+    end: usize,
+}
+
+/// Config for AnchorTxCircuit
+#[derive(Clone, Debug)]
+pub struct AnchorTxCircuitConfig<F: Field> {
+    tx_table: TxTable,
+    pi_table: PiTable,
+
+    q_tag: Selector,
+    // the anchor transaction fixed fields
+    // Gas, GasPrice, CallerAddress, CalleeAddress, IsCreate, Value, CallDataLength,
+    // 2 rows: 0, tag, 0, value
+    tag: Column<Fixed>,
+    use_rlc: Column<Fixed>,
+
+    // check: method_signature, l1Hash, l1SignalRoot, l1Height, parentGasUsed
+    q_call_data_start: Selector,
+    q_call_data_step: Selector,
+    q_call_data_end: Selector,
```

Comment:
There might be an opportunity to save 2 selectors here by using 'edge-detection' logic:

You could use one selector `q_call_data` that:
- is zero where the `q_call_data_start` and `q_call_data_end` are 1 
- is 1 in the rest of the data rows

Then in the gates you can check the "start" part by triggering when 
q_call_data at Rotation(0) = 0 AND q_call_data at Rotation(1) = 1.

Similarly for the "end" part.

Not sure though what the implications are in terms of performance.

Reply:
Seems like a good optimization. A bit less straightforward but still easy to understand.

This would increase the expression degree by 1 for constraints using these combined selectors. But if that doesn't make it so the circuit has a higher max expression degree then that's not an issue.

Maybe one other concern is that this would not work I think for single byte values, but not necessary now and also unlikely to be necessary in the future.

Reply:
Yeah, I understand this approach:

| q_call_data | call_data|
|------------|---------|
| 0 | field0[0] |
| 1 | field0[1] |
| 1 | field0[2] |
| 0 | field0[4] | 

* `field0` is `method_signature`

- `q_call_data_start` is equal `Rotation(0) == 0 ` AND `Rotation(1) = 1`
- `q_call_data_step` is equal `Rotation(0) == 1 `
- `q_call_data_end` is equal `Rotation(0) == 0 ` AND `Rotation(1) = 0`

Reply:
But we need to use Advice Column as Selector with Rotation.  Does this increase the degree?

Reply:
Yes, the zero values are tricky. As you say, the `q_call_data_end` is not easy to represent.

Reply:
Counting complexity and performance, maybe this method is not much better than the current one?
cc @Brechtpd

Reply:
Some methods also need the `Selector` type as input, like `BaseConstraintBuilder::gate`

Reply:
Simpler is definitely better in this case, so if it complicates things would indeed stick with the current method.